---
title: "Code break vignette for `recountmethylation` Bioc2021 demo"
abstract: "This vignette walks through all of the examples in the code break sessions for the [`recountmethylation` Bioc2021 demo](https://bioc2021.bioconductor.org/posts/paper106/). Covered topics include how to work with DNAm array data from the HM450K or EPIC platform as `RGChannelSet`s, how to obtain the database compilation files with `getdb` functions, how to convert between `SummarizedExperiment` and `h5se`/`HDF5-SummarizedExperiment` objects, and how to search in the harmonized metadata for samples of interest. Additional resources have been provided at the [demo's GitHub repo](https://github.com/metamaden/recountmethylation_bioc2021)."
author: Sean K. Maden, Ph.D. candidate, Oregon Health & Science University
keywords: "R, Bioconductor, SummarizedExperiment, minfi, DelayedArray, DNA methylation, epigentics, meta-analysis, Gene Expression Omnibus, BeadArray, HM450K, EPIC"
output:
  pdf_document:
    toc: true
    number_sections: true
    keep_tex: true
    fig_caption: true
    latex_engine: pdflatex
  html_document: default
date: "`r format(Sys.time(), '%B %d, %Y')`"
geometry: margin=1in
fontfamily: mathpazo
fontsize: 11pt
---

```{r setup, include=FALSE}
library(recountmethylation)
library(minfi)
library(minfiData)
library(minfiDataEPIC)
library(HDF5Array)
```

# Code break #1 -- Working with HM450K and EPIC DNAm array data as `RGChannelSet` objects

This code break will demonstrate how to work with `RGChannelSet` objects, a subclass of `SummarizedExperiment` container object. Objects will include DNAm array data from either the HM450K or EPIC DNAm array platform. Example datasets will be loaded from `minfiData` and `minfiDataEPIC`, and `minfi` functions will be used to convert and combine the data platforms.

## Install and load the dependencies

If you haven't already installed the dependencies, do this now with `BiocManager::install(c("minfi", "minfiData", "minfiDataEPIC"))`. Next, use `library` to load these:

```{r}
library(minfi)
library(minfiData)
library(minfiDataEPIC)
```

## Access the example `RGChannelSet` objects

We'll use the `data()` function to access the pre-compiled example `RGChannelSet` objects containing HM450K and EPIC DNAm array data. Next, assign these example objects to informative variables.

```{r}
data("RGsetEx") # load the example hm450k data
data("RGsetEPIC") # load the example epic data
rg.hm450k <- RGsetEx
rg.epic <- RGsetEPIC
```

Note when you are working with your own data in the form of the IDAT intensity files, you will read the data in using one of `minfi`'s functions for this task, such as `read.metharray()` or `read.metharray.exp()`.

## Inspect the `RGChannelSet` objects

View the object classes with `class()`.

```{r}
class(rg.hm450k)
class(rg.epic)
```

These are both `RGChannelSet` objects, a type of `SummarizedExperiment` container used to handle DNAm array data.

You can view the annotation strings with `annotation()`.

```{r}
annotation(rg.hm450k)
annotation(rg.epic)
```

These annotation strings will associate the contained assays with feature annotations from the manifest, which we'll access using `getAnnotation()` below.

Inspect the dimensions of the assay data matrix with `dim()`.

```{r}
dim(rg.hm450k)
dim(rg.epic)
```

Note, the first dimension returned, or the number of rows, reflects the number of BeadArray assays read from the IDATs. Because more than one such assay can inform a single array probe measurement, these rows exceed the total number of array probes. Also note the column count reflects the number of samples in the assays matrix.

Next, we'll access the green channel intensities data contained in the example objects using the `getGreen()` function. 

```{r}
gm.hm450k <- getGreen(rg.hm450k)
gm.epic <- getGreen(rg.epic)
dim(gm.hm450k)
dim(gm.epic)
class(gm.hm450k)
class(gm.epic)
```

The red channel intensities data can be similarly accessed using `getRed()`. Note the dimensions in the green color channel-specific intensities matrices returned by `getGreen()` reflect the total BeadArray assays as returned by `dim()`.

Next, view the array annotation, or manifest, with `getAnnotation()`.

```{r}
anno.hm450k <- getAnnotation(rg.hm450k)
anno.epic <- getAnnotation(rg.epic)
dim(anno.hm450k)
dim(anno.epic)
```

The number of rows in these annotations reflects the total probes on each of the array platforms.

You can view the mappings of the assay measures to the probes by accessing the columns `AddressA` and `AddressB` in the annotations. These addresses reflect the assay measures, and they can be viewed in the assays matrices using `rownames()`.

```{r}
rownames.gm <- rownames(getGreen(rg.hm450k))
head(rownames.gm)
anno.filt <- anno.hm450k$AddressA %in% rownames.gm
anno.filt <- anno.filt | anno.hm450k$AddressB %in% rownames.gm
dim(anno.hm450k[anno.filt,])
```

## Converting between platforms

When analyzing public DNAm array data, you may identify similar samples run on both the HM450K and EPIC platforms. Since 93% of HM450K probes overlap EPIC, one experimental approach is to harmonize the arrays together before analysis, to boost sample size and power.

Convert the `rg.epic` data to the HM450K platform with the `convertPlatform()` `minfi` function, and inspect its contents.

```{r}
# declare a valid annotation string
anno.str <- "IlluminaHumanMethylation450k"
# convert epic to hm450k data
rg.converted <- convertArray(rg.epic, anno.str)
# inspect rg.converted
class(rg.converted)
annotation(rg.converted)
dim(rg.converted)
```

Note the data annotation was chanaged to reflect HM450K, and that the assay dimensions, or number of rows from `dim()`, are slightly lower than those in `rg.hm450k`. This reflects the fact that not all of HM450K's probes are available on the EPIC platform.

## Combining data from different platforms

We can combine arrays from the same or different platforms with the `combineArrays()` `minfi` function. If one of the platforms doesn't match the platform specified by the `outType` argument, it will be converted prior to being combined.

Let's combine `rg.hm450k` and `rg.epic` as a single `RGChannelSet` set on the HM450K platform. We'll use the same annotation string as `anno.str` above.

```{r}
rg.combined <- combineArrays(rg.epic, rg.hm450k, outType = anno.str)
dim(rg.combined)
```

The combined `RGChannelSet` now includes a subset of assays found on the HM450K platform, and the total samples, or number of columns from `dim()` is now 9.

# Code break #2 -- Obtaining, converting, and using `h5se` objects

This code break will demonstrate how to work with `h5se`/`HDF5-SummarizedExperiment` objects, including details about available datasets to download, and how to convert from `SummarizedExperiment` and back.

## Obtaining the database compilation files with the `getdb` functions

First, load the `recountmethylation` library.

```{r}
library(recountmethylation)
```

Next, inspect the docstrings for the `getdb` functions with `?getdb`. Note there are functions corresponding to each class of database file available, including `getdb_h5se_gr` for the `h5se`/`GenomicRatioSet` data, `getdb_h5se_rg` for the `h5se`/`RGChannelSet` data, etc. 

The first argument for these functions is the platform, which can be either "hm450k" or "epic" ("hm450k" is the default). The second argument is the path to the directory to search for similarly named dataset files, and where the new download will be located. If a simliarly named file is detected, there will be prompts for whether to use the existing file, and whether to overwrite the existing file.

When the `getdb` functions run, they will display the download progress in you console. After successfully completing the download, an attempt to load the downloaded object will be made.

For this example, let's download a small example dataset.

```{r}
getdb_h5se_test(platform = "hm450k", dfp = ".")
```

Having successfully download this file, we can load and inspect it.

```{r}
h5se.fname <- "remethdb-h5se_gr-test_0-0-1_1590090412"
gr.h5se.test <- loadHDF5SummarizedExperiment(h5se.fname)
dim(gr.h5se.test)
class(getBeta(gr.h5se.test))
```

Note the assays, or rownames from `dim()`, reflect probes in `GenomicRatioSet` objects. Further, note the class of the DNAm fractions (a.k.a. "Beta-values") matrix, returned by `getBeta()`, is a type of `DelayedArray` matrix.

## Convert from `RGChannelSet` to `h5se`

We can convert a `SummarizedExperiment` object to a `DelayedArray`-backed `h5se` object with the `saveHDF5SummarizedExperiment()` function from the `HDF5Array` library.

Let's do this using the example HM450k `rg.hm450k` data now. We'll save this as an `h5se` object with the sub-directory name `rg_h5se`.

```{r}
rg <- RGsetEx
h5se.dir <- "rg_h5se"
saveHDF5SummarizedExperiment(x = rg, dir = h5se.dir)
```

Note the `h5se` object is actually a directory of two files: a larger "assays.h5" file, and a smaller "se.rds" file.

```{r}
list.files(h5se.dir)
```

To load the new `h5se` object, we point to the directory and use `HDF5Array`'s `loadHDF5SummarizedExperiment` function.

```{r}
rg.h5se <- loadHDF5SummarizedExperiment(h5se.dir)
```

Inspecting `rg.h5se` shows it's still an `RGChannelSet` object, but accessing its component assay matrices returns `DelayedArray` matrices.

```{r}
class(rg.h5se)
class(getBeta(rg.h5se))
class(getGreen(rg.h5se))
```

## Troubleshooting `h5se` downloads

Downloading the full `h5se` database files will take longer than shown in our example above. If we try to access the `h5se` data subdirectories too early (e.g. before both files have downlaoded), or if one of the requisite files is corrupted, we'll see an error. We can simulate this by removing one of the requisite files for `rg.h5se` using `unlink`.

```{r, error = T}
unlink(file.path(h5se.dir, "se.rds"))
rg.h5se <- loadHDF5SummarizedExperiment(h5se.dir)
```

Let's now resave to the `h5se.dir` data directory with both requisite files. We'll do this by passing `replace = T` to the save function, since the directory already exists.

```{r}
saveHDF5SummarizedExperiment(x = rg, dir = h5se.dir, replace = T)
```

## Rapidly update sample metadata without resaving assays

For very large `h5se` objects, you may find you want to update the stored sample metadata without overwriting all the assays data. We can access and modify the sample metadata with the `colData()` function, and then update the stored metadata using `HDF5Array`'s `quickResaveHDF5SummarizedExperiment()` function. Note you only need to pass the object in memory, and it will update the target `h5se` data directory automatically.

```{r}
coldata <- colData(rg.h5se)
colnames(coldata)
colData(rg.h5se)$newvar <- "newvalues"
quickResaveHDF5SummarizedExperiment(rg.h5se)
```

Reloading the `rg.h5se` shows the new variable `newvar` was stored and successfully reloaded.

```{r}
rg.h5se <- loadHDF5SummarizedExperiment(h5se.dir)
rg.h5se$newvar
```

## Recasting an `h5se` object as a regular `SummarizedExperiment` object

Occassionally, you will need to run a function that doesn't support the `DelayedArray` backend found in `h5se` datasets. For example, let's try to run `minfi`'s `preprocessFunnorm()` function on the `rg.h5se` dataset.

```{r, error = T}
minfi::preprocessFunnorm(rg.h5se)
```

We see that an error is thrown relating to the type of data passed to the function.

In these cases, you can recast the `h5se` dataset, or revert it back, into a normal `SummarizedExperiment` type container. When working with the full-sized `h5se` objects, you may need to recast the data in chunks to save memory. 

For this example, recast the entire `rg.h5se` dataset in one go. We'll isolate the basic `RGChannelSet` components for the `Green`, `Red`, and `annotation` slots in the new object.

```{r}
gm <- as.matrix(getGreen(rg.h5se))
rm <- as.matrix(getRed(rg.h5se))
anno <- annotation(rg.h5se)
```

Now we make a new object with `RGChannelSet()` and store this in the `rg.new` variable.

```{r}
rg.new <- RGChannelSet(Green = gm, Red = rm, annotation = anno)
```

Running the preprocess function on `rg.new` will now complete successfully.

```{r}
gr.new <- minfi::preprocessFunnorm(rg.new)
```

Inspecting `gr.new`, we see this is a `GenomicRatioSet` that contains normalized probe-level measurements.

```{r}
class(gr.new)
annotation(gr.new)
dim(gr.new)
dim(getBeta(gr.new))
```

This can be further recast as an `h5se` object using the same strategy as shown above.

```{r, echo = F, eval = T, messages = F, warnings = F}
unlink(h5se.dir, recursive = T)
unlink("remethdb-h5se_gr-test_0-0-1_1590090412", recursive = T)
```

# Code break #3 -- Querying and filtering samples of interest using harmonized metadata

This code break covers how to summarize, query, and filter the harmonized sample metadata found in `recountmethylation`'s database compilation files.

## Load a full-sized `h5se` object containing EPIC DNAm array data

This example will use an `h5se` dataset containing a recent compilation of EPIC DNAm array data with available IDATs from GEO. A version of this dataset will likely appear on the server as part of an upcoming paper.

Let's first load the `HDF5Array` library.

```{r}
library(HDF5Array)
```

Next, let's note the information in the name of the `h5se` object's data directory: "remethdb_h5se_rg_epic-hm850k_merged_1621537799-1589820348_0-0-3". We can see a lot of details about the contained data. For instance, this `h5se` object includes the `rg`/`RGChannelSet` for the "epic" platform, including data that was merged between two files with distinct timestamps, "1621537799" and "1589820348". Informative file titles can simplify your work as you produce new compilations.

With the test `h5se` data directory in the current working directory, let's load and inspect the dataset.

```{r}
h5se.fname <- "remethdb_h5se_rg_epic-hm850k_merged_1621537799-1589820348_0-0-3"
rg <- loadHDF5SummarizedExperiment(h5se.fname)
class(rg)
annotation(rg)
dim(rg)
```

We can see the assay quantity, or number of rows from `dim()`, reflects the total bead assay measures for the EPIC platform.

## Summarize the harmonized metadata

Let's now view some of the available sample metadata in this object with `colData()`.

```{r}
cd <- colData(rg)
colnames(cd)[1:10]
```

We can see there are columns for the sample/gsm ID ("gsm"), study/gse ID ("gse"), the sample title string ("gsm_title"), the mined storage info ("storageinfo"), the mined demographic variables for "sex" and "age", the learned tissue terms ("tissue"), the learned disease and experiment condition terms ("disease"), the most likely sample type prediction from the MetaSRA-pipeline ("sampletype"), and the predicted age ("predage").

Let's inspect the non-NA learned terms under the "tissue" variable.

```{r}
head(cd$tissue[!cd$tissue == "NA"])
```

Here, we can see the uniform formatting for the learned terms: underscores instead of spaces; all lower case letters; and terms separated by semicolons. Also note the extent of detail in these learned terms varies considerably by sample.

Let's now view the most frequent terms under "tissue". We can get the vector of all unique learned terms using `unlist` on `strsplit` where we split terms on their separating semicolons. Finally, we coerce a frequency table to a data frame and sort it.

```{r}
tissuev <- unlist(strsplit(cd$tissue, ";"))
df.tissue <- as.data.frame(table(tissuev))
df.tissue <- df.tissue[rev(order(df.tissue[,2])),]
head(df.tissue)
```

Viewing the top most frequent terms under "tissue", we see these mainly relate to blood and cancer samples.

## Query the metadata for colon samples

Next, let's use regular expressions to precisely match a query. Here, we'll search for samples with the learned term "colon". We'll ensure this term occurs in its entirety, either by itself or embedded with other terms flanked by ";". Thus, we use the regular expression pattern "(^|;)colon($|;)" in the `grepl()` function for this search.

```{r}
tx.query.str <- "colon"
tx.query.str.format <- paste0("(^|;)",tx.query.str,"($|;)")
tx.filt <- grepl(tx.query.str.format, cd$tissue)
cdf <- cd[tx.filt,]
dim(cdf)
```

We see that 321 sample records remain after filtering on "colon" under the tissue variable. Next, let's summarize the studies and samples per study in this filtered sample set.

```{r}
df.gse <- as.data.frame(table(cdf$gse))
df.gse <- df.gse[rev(order(df.gse[,2])),]
df.gse
dim(df.gse)
```

We see there are 9 total unique study records, or gse IDs, which contribute between 4 and 206 samples.

## Filter studies on their sample counts

We can now apply a minimum samples filter. This can be useful if you intend to perform a study-wise bias correction, and want to ensure enough samples are contributed to yield an accurate study bias estimate. 

Let's now use a filter to retain samples from studies having at least 20 colon samples.

```{r}
min.gsm <- 20
dff.gse <- df.gse[df.gse[,2] >= min.gsm,]
dff.gse
sum(dff.gse[,2])
cdff <- cdf[cdf$gse %in% dff.gse[,1],]
dim(cdff)
```

We're left with 3 studies and 275 samples. Let's now summarize terms learned for the "disease" variable for this subset of colon samples.

```{r}
dx.df <- as.data.frame(table(unlist(strsplit(cdff$disease, ";"))))
dx.df <- dx.df[rev(order(dx.df[,2])),]
dx.df
```

We can see terms for cancer and colorectal cancer, indicating that some of these samples may be from cancer patients or tumor samples. 

## Annotate studies and samples by consulting the GEO records

We can now start to clarify the types of colon samples we've identified. A reasonable and time-effective way to do this is to inspect the study records in GEO from your browser, and to begin a custom annotation of the studies themselves. This is more rapid than reviewing all samples individually, as you'll often find characteristics at the study level which also apply to each of its samples.

An example record for the study with gse ID GSE132804 is:

![GSE record for GSE132804](gse_record.jpg)



From the record, we can note this was a study of dysfuntional epigenetic aging in normal colon samples from patients with low to high colorectal cancer risk, based on their clinical record and history of cancer or adenoma (full disclosure: I was the second author on this study :^] ). 

If we're studying normal colon, we may or may not care if the sample came from a patient with current or prior cancer, or current or prior adenoma. These factors can be important for deciding on samples to include and exclude as you analyze the publicly available datasets.

An example study-level annotation is shown:

```{r}
lgse <- list()
lgse[["GSE132804"]] <- list(title = "Dysfunctional epigenetic aging of the normal colon and colorectal cancer risk", 
                            notes = "Subjects include low to high CRC risk, with possible adenoma or CRC history.")
lgse[["GSE149282"]] <- list(title = "Genome-wide open chromatin methylome profiles in colorectal cancer", 
                            notes = "Study in CRC patients, includes cancer tissues and normal tissues adjacent to cancers.")
lgse[["GSE166212"]] <- list(title = "Colorectal Cancer DNA Methylation", 
                            notes = "Study in CRC tumors, includes demographics and tumor type info.")
```

## Compare a sample's harmonized metadata with its online record in GEO

With GEO opened in a browser, we can also review a selection of samples and check the learned and harmonized metadata against data available in the online record. Notice for the sample with gsm ID GSM5065983, the GEO record looks like:

![Example GSM record for GSM5065983](gsm_record.jpg)

```{r}
gsm.row <- cdff[cdff$gsm == "GSM5065983",]
gsm.row$tissue
gsm.row$disease
gsm.row$sex
gsm.row$age
```

Comparing the learned versus stored metadata, we see terms for sex and age appear accurate, and there's an "NA" value for "age_info", reflecting that the numeric age units couldn't be mined for this sample. Learned terms in the "tissue" and "disease" fields further convey this is a colorectal cancer sample, although whether it precisely originates from colon or rectum is unclear.

## Filter samples according to the presence of the "cancer" term

Finally, we can reuse the above approach exclude samples with "cancer" learned under the "tissue" variable. This leaves us with 223 colon samples.

```{r}
tx.normal.filt <- !grepl("(^|;)cancer($|;)", cdff$tissue)
cdfff <- cdff[tx.normal.filt,]
dim(cdfff)
```

# Next steps

For additional help with analysis of DNAm array data in the R environment with Bioconductor libraries, you can consult the documentation for the [`minfi`](https://bioconductor.org/packages/release/bioc/html/minfi.html), [`HDF5Array`](https://bioconductor.org/packages/release/bioc/html/HDF5Array.html), [`DelayedArray`](https://bioconductor.org/packages/release/bioc/html/DelayedArray.html), and [`recountmethylation`](https://bioconductor.org/packages/release/bioc/html/recountmethylation.html) libraries. Additional resources have been provided in the wiki at the [GitHub repo](https://github.com/metamaden/recountmethylation_bioc2021) for this software demo.

# Citations

If you use `recountmethylation` in published work, please cite the following paper:

Sean K Maden, Reid F Thompson, Kasper D Hansen, Abhinav Nellore, Human methylome variation across Infinium 450K data on the Gene Expression Omnibus, NAR Genomics and Bioinformatics, Volume 3, Issue 2, June 2021, lqab025, https://doi.org/10.1093/nargab/lqab025

This document uses the pdf template available at [github.com/svmiller/svm-r-markdown-templates/](https://github.com/svmiller/svm-r-markdown-templates/tree/master/article-example) 

Credit is further due to the authors of the Bioconductor libraries featured above, including `minfi` and `HDF5Array`. Their citations are viewable using the `citation()` function.

```{r}
citation("minfi")
citation("HDF5Array")
```

# Session info

The session info shows the versions of R and various packages used in this vignette. Showing this is an important part of reproducible research. View this as follows:

```{r}
sessionInfo()
```
